### PR TITLE
test: support storage tests with non-standard kubelet root directory

### LINKED ISSUE #92664: CSI Volume tests should work with non-default kubelet root directories
**What happened**:
The CSI hostpath driver used in the e2e tests expects the root directory of nodes to be set to `/var/lib/kubelet`, [hardcoded here](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/testing-manifests/storage-csi/hostpath/hostpath/csi-hostpath-plugin.yaml#L130-L133), however nodes can be configured to use different paths. The tests should find out the root directory from kubelet to allow for this.

Several test failures related to CSI Volumes:
```
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)(allowExpansion)] volume-expand [Top Level] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)(allowExpansion)] volume-expand Verify if offline PVC expansion works [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)(allowExpansion)] volume-expand [Top Level] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)(allowExpansion)] volume-expand should resize volume when PVC is edited while pod is using it [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)] volume-expand [Top Level] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)] volume-expand should not allow expansion of pvcs without AllowVolumeExpansion property [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)] volumeMode [Top Level] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)] volumeMode should not mount / map unused volumes in a pod [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)] volumes [Top Level] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)] volumes should store data [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)(allowExpansion)] volume-expand [Top Level] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)(allowExpansion)] volume-expand Verify if offline PVC expansion works [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)(allowExpansion)] volume-expand [Top Level] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)(allowExpansion)] volume-expand should resize volume when PVC is edited while pod is using it [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] provisioning [Top Level] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with pvc data source [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] provisioning [Top Level] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with snapshot data source [Feature:VolumeSnapshotDataSource] [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath [Top Level] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath should support existing directories when readOnly specified in the volumeSource [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] volumes [Top Level] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] volumes should store data [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic Snapshot] snapshottable [Top Level] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic Snapshot] snapshottable should create snapshot with defaults [Feature:VolumeSnapshotDataSource] [Suite:openshift/conformance/parallel] [Suite:k8s]
```

**What you expected to happen**: Test should be expected to pass.

**How to reproduce it (as minimally and precisely as possible)**:

- Configure kubelet on cluster nodes to use the non-default path.
- Run the storage-csi e2e tests.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
$ oc version
Client Version: 4.3.3
Server Version: 4.4.6
Kubernetes Version: v1.17.1+f63db30
```
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Some storage tests deploy DaemonSets which hard-code /var/lib/kubelet as root
directory for kubelet registration and pod directory. To enable those tests in
clusters with some other directory, a new global parameter gets added and the
value defined there gets patched into the relevant places of a CSI driver's
YAML files.

#### Which issue(s) this PR fixes:

Fixes #92664

#### Special notes for your reviewer:

A global parameter is used because other tests might also need the same
information.

There are currently no Prow jobs which use a non-standard path. I verified manually that the new flag has the intended behavior:
```
# Without flags.
$ kubectl get pods --all-namespaces -o yaml | grep /var/lib/kubelet
                k:{"mountPath":"/var/lib/kubelet/plugins"}:
                k:{"mountPath":"/var/lib/kubelet/pods"}:
      - mountPath: /var/lib/kubelet/pods
      - mountPath: /var/lib/kubelet/plugins
      - --kubelet-registration-path=/var/lib/kubelet/plugins/csi-hostpath-provisioning-8787/csi.sock
        path: /var/lib/kubelet/plugins/csi-hostpath-provisioning-8787
        path: /var/lib/kubelet/pods
        path: /var/lib/kubelet/plugins_registry
        path: /var/lib/kubelet/plugins

# With -kubelet-root=/var/lib/kubeletxxx
$ kubectl get pods --all-namespaces -o yaml | grep /var/lib/kubelet
                k:{"mountPath":"/var/lib/kubeletxxx/plugins"}:
                k:{"mountPath":"/var/lib/kubeletxxx/pods"}:
      - mountPath: /var/lib/kubeletxxx/pods
      - mountPath: /var/lib/kubeletxxx/plugins
      - --kubelet-registration-path=/var/lib/kubeletxxx/plugins/csi-hostpath-provisioning-4758/csi.sock
        path: /var/lib/kubeletxxx/plugins/csi-hostpath-provisioning-4758
        path: /var/lib/kubeletxxx/pods
        path: /var/lib/kubeletxxx/plugins_registry
        path: /var/lib/kubeletxxx/plugins
```

#### Does this PR introduce a user-facing change?

```release-note
The e2e.test binary supports a new `--kubelet-root` parameter to override the default `/var/lib/kubelet` path. CSI storage tests use this.
```

### CODE REVIEW

**Comment by pohly** [comment]
/retest

/hold

I'm working on a Prow job with non-standard kubelet root. Let's see whether I can get that working before merging this PR.

**Comment by pohly** [comment]
The idea with using a unique path in pull-kubernetes-e2e-kind and then using that path to find leaked loop devices [didn't work out](https://kubernetes.slack.com/archives/CEKK1KTN2/p1645554213273549?thread_ts=1645458120.762849&cid=CEKK1KTN2). Therefore running pull-kubernetes-e2e-kind with a non-standard path is probably not worth it. But that is debatable: on the one hand it ensures that tests don't make assumptions about the kubelet path, on the other hand it creates additional friction for the upstream maintainers.

This PR currently also doesn't change where csi-driver-host-path puts its persistent state. That is still /var/lib/csi-data-dir.  @jonesbr17: do you know whether that might be a problem in the cluster where you encountered issue https://github.com/kubernetes/kubernetes/issues/92664?

For future reference, here's a patch which also changes csi-data-dir:
```diff
diff --git a/test/e2e/storage/drivers/csi.go b/test/e2e/storage/drivers/csi.go
index 10af124fc4c..9ca76bb0cfa 100644
--- a/test/e2e/storage/drivers/csi.go
+++ b/test/e2e/storage/drivers/csi.go
@@ -45,6 +45,7 @@ import (
 	"strings"
 	"sync"
 	"time"
+	"path"
 
 	"github.com/onsi/ginkgo"
 	spb "google.golang.org/genproto/googleapis/rpc/status"
@@ -73,6 +74,7 @@ import (
 	"k8s.io/kubernetes/test/e2e/storage/drivers/proxy"
 	storageframework "k8s.io/kubernetes/test/e2e/storage/framework"
 	"k8s.io/kubernetes/test/e2e/storage/utils"
+	e2eframework "k8s.io/kubernetes/test/e2e/framework"
 
 	"google.golang.org/grpc"
 )
@@ -227,6 +229,20 @@ func (h *hostpathCSIDriver) PrepareTest(f *framework.Framework) (*storageframewo
 		DriverNamespace:     driverNamespace,
 	}
 
+	// We put the state directory alongside the kubelet state directory:
+	// /var/lib/kubelet -> /var/lib/csi-hostpath-data.
+	//
+	// If kubelet is told to use something like /var/lib/<test specific ID>/kubelet,
+	// then our directory also automatically uses that test specific directory.
+	// We also make that available inside the container under the same path as on
+	// the host, in contrast to the default deployment file, which uses /csi-data-dir.
+	//
+	// This is relevant for loop device cleanup under kind Prow jobs: when a loop
+	// device shows up with /var/lib/<test specific ID> after the cluster has been
+	// torn down, https://github.com/kubernetes-sigs/kind/blob/main/hack/ci/e2e-k8s.sh
+	// knows that it can delete the loop device.
+	statedir := path.Join(path.Dir(strings.TrimRight(e2eframework.TestContext.KubeletRoot, "/")), "csi-hostpath-data")
+
 	o := utils.PatchCSIOptions{
 		OldDriverName:       h.driverInfo.Name,
 		NewDriverName:       config.GetUniqueDriverName(),
@@ -239,6 +255,8 @@ func (h *hostpathCSIDriver) PrepareTest(f *framework.Framework) (*storageframewo
 			// Enable volume lifecycle checks, to report failure if
 			// the volume is not unpublished / unstaged correctly.
 			"--check-volume-lifecycle=true",
+
+			"--statedir", statedir,
 		},
 		ProvisionerContainerName: "csi-provisioner",
 		SnapshotterContainerName: "csi-snapshotter",
@@ -272,12 +290,28 @@ func (h *hostpathCSIDriver) PrepareTest(f *framework.Framework) (*storageframewo
 				switch container.Name {
 				case "csi-external-health-monitor-agent", "csi-external-health-monitor-controller":
 					// Remove these containers.
+				case "hostpath":
+					for i, volmount := range container.VolumeMounts {
+						if volmount.Name == "csi-data-dir" {
+							volmount.MountPath = statedir
+							container.VolumeMounts[i] = volmount
+							break
+						}
+					}
+					containers = append(containers, container)
 				default:
 					// Keep the others.
 					containers = append(containers, container)
 				}
 			}
 			item.Spec.Template.Spec.Containers = containers
+			for i, volume := range item.Spec.Template.Spec.Volumes {
+				if volume.Name == "csi-data-dir" {
+					volume.HostPath.Path = statedir
+					item.Spec.Template.Spec.Volumes[i] = volume
+					break
+				}
+			}
 		}
 		return nil
 	}, h.manifests...)
```

I tested this PR here locally with a modified version of https://github.com/kubernetes-sigs/kind/blob/main/hack/ci/e2e-k8s.sh:

```diff
diff --git a/hack/ci/e2e-k8s.sh b/hack/ci/e2e-k8s.sh
index 9d91b21c..c42bbd3c 100755
--- a/hack/ci/e2e-k8s.sh
+++ b/hack/ci/e2e-k8s.sh
@@ -26,6 +26,30 @@ set -o errexit -o nounset -o xtrace
 #          false - (default) APIs and features left at defaults
 # 
 
+# This unique ID will be used to run kubelet with a --root-dir that is specific
+# to this test run. Loop devices created for raw block volumes will have a path
+# that is inside that directory and will show up on the host with that path (no
+# namespacing for loop devices!). While cleaning up we can delete those.
+#
+# Only some Kubernetes test/e2e binaries have support for running with a
+# non-standard kubelet root directory (added in
+# https://github.com/kubernetes/kubernetes/pull/108253). We leave this empty
+# if not available and skip loop device cleanup.
+TEST_UUID=
+TEST_ROOT_DIR=
+KUBELET_ROOT_DIR=
+if grep -q kubelet-root test/e2e/framework/test_context.go; then
+    TEST_UUID=$(dd if=/dev/random count=16 bs=1 status=none | base64)
+    TEST_ROOT_DIR="/var/lib/$TEST_UUID"
+    KUBELET_ROOT_DIR="${TEST_ROOT_DIR}/kubelet"
+fi
+
+# Root privileges are required for cleaning up loop devices.
+SUDO=
+if [ $(id -u) -ne 0 ]; then
+    SUDO=sudo
+fi
+
 # cleanup logic for cleanup on exit
 CLEANED_UP=false
 cleanup() {
@@ -42,6 +66,17 @@ cleanup() {
   if [ -n "${TMP_DIR:-}" ]; then
     rm -rf "${TMP_DIR:?}"
   fi
+
+  # Clean up leaked loop devices?
+  if [ "$TEST_UUID" ]; then
+    ${SUDO} losetup --noheadings --output NAME,BACK-FILE | while read -r dev path; do
+      if echo "$path" | grep -q "^${TEST_ROOT_DIR}/"; then
+        echo "Cleaning up leaked ${dev} for ${path}."
+        ${SUDO} losetup -d "${dev}"
+      fi
+    done
+  fi
+
   CLEANED_UP=true
 }
 
@@ -86,6 +121,10 @@ create_cluster() {
     esac
   fi
 
+  # NOTE: the indendation on the next line is meaningful!
+  kubelet_extra_args="${kubelet_extra_args}
+      \"root-dir\": \"${KUBELET_ROOT_DIR}\""
+
   # JSON map injected into featureGates config
   feature_gates="{}"
   # --runtime-config argument value passed to the API server
@@ -229,6 +268,7 @@ run_tests() {
   # running a builtin like `wait`, saving the PID also allows us to forward the
   # interrupt
   ./hack/ginkgo-e2e.sh \
+    $(if [ "${KUBELET_ROOT_DIR}" ]; then echo "--kubelet-root=${KUBELET_ROOT_DIR}"; fi) \
     '--provider=skeleton' "--num-nodes=${NUM_NODES}" \
     "--ginkgo.focus=${FOCUS}" "--ginkgo.skip=${SKIP}" \
     "--report-dir=${ARTIFACTS}" '--disable-log-dump=true' &
```

Tests passed, just the loop device cleanup didn't work out as intended.

/hold cancel

**Comment by pacoxu** [comment]
It is a little dup with `KubeVolumeDir`.

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/test_context.go#L340

**Comment by pohly** [comment]
> It is a little dup with KubeVolumeDir.

Good catch, I hadn't seen that one.

> rename it to KubeRootDir and use for both scenarios. (But it may affect some volume test cases.) Or can we reuse it directly?

You want me to rename `KubeletRoot` to `KubeRootDir`, right? `KubeletRootDir` makes sense to me, but not so much `KubeRootDir` - what would `Kube` stand for, i.e. which component of Kubernetes?

I'm not sure what you mean with `Or can we reuse it directly?` - what is `it` here? `KubeVolumeDir`? Using that and going up one level feels hacky to me.

**Comment by pacoxu** [comment]
I mean can we reuse `KubeVolumeDir`?  If the name is not valid for this purpose, we may rename it.

**Comment by pohly** [comment]
`KubeVolumeDir` is too specific. We can of course assume that `KubeletRootDir = KubeVolumeDir/..` but that looks backwards to me. I prefer to specify `KubeletRootDir` and then derived the other directories from that.

**Comment by pacoxu** [comment]
> I prefer to specify `KubeletRootDir` and then derived the other directories from that.

I agree.

**Comment by pohly** [comment]
I must have been confused by the `KubeVolumeDir` name: that parameter already *is* the kubelet root dir, not the `pods/*/volumes` directory underneath that.

I think the parameter was badly named when it was introduced. I'm now torn between renaming the parameter (may cause problems for anyone already using it) and just updating the parameter's description (less intrusive, my still harder to discover).

@pacoxu : any preference?

**Comment by pohly** [comment]
There's a compromise: add the new parameter and rename the field, but keep `--volume-dir` as an alias. I've implemented that approach.

**Comment by pohly** [comment]
@pacoxu : please take another look, see https://github.com/kubernetes/kubernetes/compare/d7b0eb25efad866e91b0a4e309fe877c6d5f7916..ff5558edc33f6042171ac3034e87e08926dab1f5 for changes.

**Review by pacoxu — APPROVED** [review]
/lgtm

**Comment by pohly** [comment]
@SergeyKanzhelev @ehashman: there are some comments in node tests about "framework.TestContext.KubeVolumeDir is currently not populated for node e2e", for example here:

https://github.com/kubernetes/kubernetes/blob/5b9910a574340e111a6776bdfd20c6a36a538874/test/e2e_node/node_problem_detector_linux.go#L297-L299

I don't know what is meant with "not populated" and therefore didn't touch those parts.

The default `/var/lib/kubelet` should always be set, so using it would be better than hard-coding (in that example) `hostLogFile = "/var/lib/kubelet/pods/"`. Let me know whether you want me update also those code lines.

**Review by ehashman — COMMENTED** [review]
/approve

for node test

**Pull request merged**
